modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/config/ParsedConfigs.scala

```diff
@@ -111,31 +115,45 @@ object ParsedConfigs {
           }
       case OutputConfig.Kinesis(_, _, Some(key), _, _, _) if !enrichedFieldsMap.contains(key) =>
         NonEmptyList.one(s"Partition key $key not valid").invalid
+      case ka: OutputConfig.Kafka if !ka.headers.forall(enrichedFieldsMap.contains) =>
+        NonEmptyList
+          .one(
+            s"Fields [${ka.headers.filterNot(enrichedFieldsMap.contains).mkString(", ")}] for headers are not part of the enriched event"
+          )
+          .invalid
       case _ =>
         output.valid
     }
 
   private[config] def outputAttributes(output: OutputConfig): EnrichedEvent => Map[String, String] =
     output match {
-      case OutputConfig.PubSub(_, Some(attributes), _, _, _) =>
-        val fields = ParsedConfigs.enrichedFieldsMap.filter {
-          case (s, _) =>
-            attributes.contains(s)
-        }
-        attributesFromFields(fields)
-      case OutputConfig.Kinesis(_, _, Some(key), _, _, _) =>
-        val fields = ParsedConfigs.enrichedFieldsMap.filter {
-          case (s, _) =>
-            s == key
-        }
-        attributesFromFields(fields)
-      case _ =>
-        _ => Map.empty
+      case OutputConfig.PubSub(_, Some(attributes), _, _, _) => attributesFromFields(attributes)
+      case OutputConfig.Kinesis(_, _, Some(key), _, _, _) => attributesFromFields(Set(key))
+      case OutputConfig.Kafka(_, _, _, headers, _) => attributesFromFields(headers)
+      case _ => _ => Map.empty
+    }
+
+  private[config] def attributesFromFields(attributes: Set[String]): EnrichedEvent => Map[String, String] = {
+    val fields = ParsedConfigs.enrichedFieldsMap.filter {
+      case (s, _) =>
+        attributes.contains(s)
+    }
+    (ee: EnrichedEvent) =>
+      fields.flatMap {
+        case (k, f) =>
+          Option(f.get(ee)).map(v => k -> v.toString)
+      }
+  }
+
+  private[config] def outputPartitionKey(output: OutputConfig): EnrichedEvent => String =
+    output match {
+      case OutputConfig.Kafka(_, _, partitionKey, _, _) => partitionKeyFromFields(partitionKey)
+      case OutputConfig.Kinesis(_, _, Some(partitionKey), _, _, _) => partitionKeyFromFields(partitionKey)
+      case _ => _ => ""
     }
 
-  private def attributesFromFields(fields: Map[String, Field])(ee: EnrichedEvent): Map[String, String] =
-    fields.flatMap {
-      case (k, f) =>
-        Option(f.get(ee)).map(v => k -> v.toString)
+  private[config] def partitionKeyFromFields(partitionKey: String): EnrichedEvent => String =
+    ParsedConfigs.enrichedFieldsMap.get(partitionKey).fold[EnrichedEvent => String](_ => "") { f => ee =>
+      Option(f.get(ee)).fold("")(_.toString)
```

Comment:
Interesting here that you return `""` if the event is missing the field.  Contrast with what happens in enrich-kinesis where [a random UUID is used as the partition key](https://github.com/snowplow/enrich/blob/3.2.3/modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala#L175) if a fallback is needed.  I guess the point of the UUID is to prevent a "hot" shard that could receive a disproportionate fraction of the events.

Reply:
From kafka [docs](https://kafka.apache.org/0102/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html): 
`If neither key nor partition is present a partition will be assigned in a round-robin fashion. `

so we don't need a random UUID as we don't specify key and partition

Reply:
But you are setting a partition key.  You're setting it as `""`, which I believe Kafka treats as a valid partition key.  This is not the same thing as not setting a partition key.

I'm only like 80% sure about this, so please call me out if I'm wrong!

Reply:
I think you're right! the same doc explains that

`If no partition is specified but a key is present a partition will be chosen using a hash of the key.`

so we'll be hitting the same partition by hashing empty string -- very important catch, thanks! I'll use a random UUID here